packages/react-storage/src/components/StorageBrowser/Views/LocationActionView/useHandleUpload.ts

```diff
@@ -129,60 +138,76 @@ export function useHandleUpload({
-        const { bucket: bucketName, credentialsProvider, region } = getConfig();
-        const input: UploadDataWithPathInput = {
-          path: `${prefix}${key}`,
-          data,
-          options: {
-            bucket: { bucketName, region },
-            locationCredentialsProvider: credentialsProvider,
-            onProgress: ({ totalBytes, transferredBytes }) => {
-              const progress = totalBytes ? transferredBytes / totalBytes : 0;
-              const nextTask: TaskUpdate = {
-                key,
-                progress,
-                ...(totalBytes === transferredBytes
-                  ? { cancel: undefined }
-                  : undefined),
-              };
-              setTasks((curr) => updateTasks(curr, nextTask));
-            },
-            preventOverwrite,
+  const handleUpload = async () => {
+    let currentIndex = 0;
+
+    const uploadNext = async () => {
+      if (currentIndex >= tasks.length) {
+        return;
+      }
+
+      const task = tasks[currentIndex];
+
+      currentIndex++;
+
+      const { key, data } = task;
+      const { bucket: bucketName, credentialsProvider, region } = getConfig();
+      const input: UploadDataWithPathInput = {
+        path: `${prefix}${key}`,
+        data,
+        options: {
+          bucket: { bucketName, region },
+          locationCredentialsProvider: credentialsProvider,
+          onProgress: ({ totalBytes, transferredBytes }) => {
+            const progress = totalBytes ? transferredBytes / totalBytes : 0;
+            const nextTask: TaskUpdate = {
+              key,
+              progress,
+              ...(totalBytes === transferredBytes
+                ? { cancel: undefined }
+                : undefined),
+            };
+            setTasks((curr) => updateTasks(curr, nextTask));
           },
-        };
-
-        const { cancel, result } = uploadData(input);
-
-        result
-          .then(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'COMPLETE' }));
-          })
-          .catch(() => {
-            setTasks((curr) => updateTasks(curr, { key, status: 'FAILED' }));
-          });
-
-        const handleCancel = () => {
-          cancel();
-
+          preventOverwrite,
+        },
+      };
+
+      const { cancel, result } = uploadData(input);
+
+      task.cancel =
+        task.size >= MULTIPART_UPLOAD_THRESHOLD_BYTES
+          ? () => cancel()
+          : undefined;
+
+      setTasks((curr) =>
+        updateTasks(curr, { key, status: 'PENDING', cancel: task.cancel })
+      );
+
+      try {
+        await result;
+        setTasks((curr) => updateTasks(curr, { key, status: 'COMPLETE' }));
+      } catch (error) {
+        if (isCancelError(error)) {
           setTasks((curr) =>
-            // assign cancel to noop
-            updateTasks(curr, { key, cancel: () => null, status: 'CANCELED' })
+            updateTasks(curr, { key, status: 'CANCELED', cancel: undefined })
           );
```

Comment:
calling `result.cancel()` from `uploadData(input)` will cause the upload to throw an error and we have to check if it's a "cancel" error so that we can handle it as a cancel